roles/download/tasks/download_container.yml

```diff
@@ -1,40 +1,125 @@
-    - download.container
+    - download_force_cache
+
+- name: download_container | Set fact indicating if image is in cache
+  set_fact:
+    image_is_cached: "{{ cache_image.stat.exists | default(false) }}"
   tags:
     - facts
 
-# FIXME(mattymo): In Ansible 2.4 omitting download delegate is broken. Move back
-#                 to one task in the future.
-- name: container_download | Download containers if pull is required or told to always pull (delegate)
-  command: "{{ docker_bin_dir }}/docker pull {{ pull_args }}"
-  register: pull_task_result
-  until: pull_task_result is succeeded
+- name: download_container | Upload image to node if it is cached
+  synchronize:
+    src: "{{ image_path_cached }}"
+    dest: "{{ image_path_final }}"
+    use_ssh_args: "{{ has_bastion | default(false) }}"
+    mode: push
+  delegate_facts: no
+  register: upload_image
+  failed_when: not upload_image
+  run_once: "{{ download_run_once }}"
+  until: upload_image is succeeded
   retries: 4
   delay: "{{ retry_stagger | random + 3 }}"
-  changed_when: not 'up to date' in pull_task_result.stdout
   when:
-    - download_run_once
-    - download.enabled
-    - download.container
-    - any_pull_required | default(download_always_pull)
-  delegate_to: "{{ download_delegate }}"
-  delegate_facts: yes
-  run_once: yes
+    - download_force_cache
+    - image_is_cached
+    - not download_localhost
+    - ansible_os_family not in ["CoreOS", "Container Linux by CoreOS"]
+  tags:
+    - upload
+    - upgrade
+
+# NOTE: If download_delegate==localhost, then the account running ansible must be in the group 'docker' on the ansible host
+- name: download_container | Load image into docker
+  shell: "{{ docker_bin_dir }}/docker load < {{ image_path_cached if download_localhost else image_path_final }}"
+  delegate_to: "{{ download_delegate if download_run_once or inventory_hostname }}"
+  run_once: "{{ download_run_once }}"
+  register: container_load_status
+  failed_when: container_load_status | failed
+  become: "{{ not download_localhost }}"
+  when:
+    - image_is_cached
+    - download_force_cache
+    - ansible_os_family not in ["CoreOS", "Container Linux by CoreOS"]
+  tags:
+    - upload
+    - upgrade
+
+- name: download_container | Prepare container download
+  import_tasks: check_pull_required.yml
+  run_once: "{{ download_run_once }}"
+  when:
+    - not download_always_pull
 
-- name: container_download | Download containers if pull is required or told to always pull (all nodes)
-  command: "{{ docker_bin_dir }}/docker pull {{ pull_args }}"
+- debug:
+    msg: "XXX Pull required is: {{ pull_required }}"
+
+# NOTE: Pre-loading docker images will not prevent 'docker pull' from re-downloading the layers in that image
+# if a pull is forced. This is a known issue with docker. See https://github.com/moby/moby/issues/23684
+- name: download_container | Download image if required
+  command: "{{ docker_bin_dir }}/docker pull {{ image_reponame }}"
```

Comment:
curious, can't we use the docker module ? wouldn't that help with the note you made ?

Reply:
I'd have to look into that, as I do not know how that module works. But more so: It would be better to do that in a separate PR, as it involves implementing a new module. The current PR uses the same system that is already in place, namely the "command: docker_bin/docker foo" construct.

Reply:
fair point